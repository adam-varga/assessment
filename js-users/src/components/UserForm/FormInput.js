import React from "react";
import {
  Input,
  InputContainer,
  InputError,
  InputLabel,
} from "./UserForm.styles";
import PropTypes from "prop-types";

export default function FormInput({
  attribute,
  label,
  onChange,
  value,
  disabled,
  error,
  placeholder,
}) {
  return (
    <InputContainer>
      <InputLabel>
        {label}
        <Input
          value={value}
          onChange={(e) => onChange(attribute, e.currentTarget.value)}
          hasError={Boolean(error)}
          placeholder={placeholder}
          disabled={disabled}
        ></Input>
      </InputLabel>
      <InputError>{error}</InputError>
    </InputContainer>
  );
}

FormInput.propTypes = {
  attribute: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.arrayOf(PropTypes.string),
};
