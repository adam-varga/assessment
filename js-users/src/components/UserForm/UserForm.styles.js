import styled, { css } from "styled-components";
import { Button } from "../common.styles";

export const Backdrop = styled.div`
  position: fixed;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Popover = styled.div`
  width: 500px;
  background: white;
  border: 1px solid black;
  padding: 50px;
`;

export const Title = styled.h2`
  margin: 0 0 20px;
`;

export const InputContainer = styled.div``;

export const InputLabel = styled.label`
  display: block;
  font-size: 10px;
`;

export const Input = styled.input`
  display: block;
  border: 1px solid black;
  ${(props) =>
    props.hasError &&
    css`
      border-color: red;
    `}
`;

export const InputError = styled.p`
  margin-top: 5px;
  color: red;
  font-size: 10px;
`;

export const ButtonContainer = styled.div`
  text-align: right;
`;

export const SaveButton = styled(Button)`
  margin-left: 10px;
`;

export const CancelButton = styled(Button)`
  border-color: #555;
  color: #555;
`;
