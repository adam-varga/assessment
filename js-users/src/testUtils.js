import { createStore } from "redux";
import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import reducer from "./store/reducers";

export function renderWithRedux(
  ui,
  { initialState, store = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}
