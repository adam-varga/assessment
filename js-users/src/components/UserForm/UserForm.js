import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import {
  createAction,
  GET_USERS__REQUEST,
  UPDATE_USER__REQUEST,
  CREATE_USER__REQUEST,
  UPDATE_USER__RESET,
  CREATE_USER__RESET,
} from "../../store/actions";
import {
  Backdrop,
  Popover,
  Title,
  ButtonContainer,
  SaveButton,
  Inputs,
  CancelButton,
} from "./UserForm.styles";
import { FiLoader, FiSave, FiCheck, FiX } from "react-icons/fi";
import FormInput from "./FormInput";
import get from "lodash.get";
import { USER_STATUSES, REQUEST_STATUSES } from "../../constants";

const MODES = {
  EDIT: "EDIT",
  CREATE: "CREATE",
};

const saveButtonIcons = {
  default: FiSave,
  [REQUEST_STATUSES.PENDING]: FiLoader,
  [REQUEST_STATUSES.SUCCESS]: FiCheck,
};

class UserForm extends React.Component {
  state = {
    first_name: "",
    last_name: "",
  };

  componentDidMount() {
    this.props.resetSaveRequest();

    const { id, userRequestStatus, user } = this.props;

    if (id && userRequestStatus === REQUEST_STATUSES.SUCCESS && user) {
      this.setState({
        first_name: user.first_name,
        last_name: user.last_name,
      });
    }
  }

  componentDidUpdate(prevProps) {
    const { id, userRequestStatus, user, saveRequestStatus } = this.props;

    if (
      id &&
      userRequestStatus === REQUEST_STATUSES.SUCCESS &&
      prevProps.userRequestStatus === REQUEST_STATUSES.PENDING &&
      user
    ) {
      this.setState({
        first_name: user.first_name,
        last_name: user.last_name,
      });
    }

    if (
      saveRequestStatus === REQUEST_STATUSES.SUCCESS &&
      prevProps.saveRequestStatus === REQUEST_STATUSES.PENDING
    ) {
      setTimeout(this.close, 1500);
    }
  }

  close = () => {
    this.props.resetSaveRequest();
    this.props.close();
  };

  createUser = () => {
    this.props.createUser({
      ...this.state,
      status: USER_STATUSES.ACTIVE,
    });
  };

  updateUser = () => {
    this.props.updateUser({
      ...this.props.user,
      ...this.state,
    });
  };

  onInputChange = (attribute, value) => {
    this.setState({
      [attribute]: value,
    });
  };

  render() {
    const { id, saveRequestStatus, userRequestStatus, errors } = this.props;
    const mode = Boolean(id) ? MODES.EDIT : MODES.CREATE;

    const isLoading = id && userRequestStatus === REQUEST_STATUSES.PENDING;
    const formDisabled =
      saveRequestStatus === REQUEST_STATUSES.PENDING ||
      saveRequestStatus === REQUEST_STATUSES.SUCCESS;

    const SaveButtonIcon =
      saveButtonIcons[saveRequestStatus] || saveButtonIcons["default"];

    return (
      <Backdrop onClick={this.close}>
        <Popover onClick={(e) => e.stopPropagation()}>
          <Title>{mode === MODES.EDIT ? "Edit user" : "Create user"}</Title>
          {isLoading && "Loading..."}
          {!isLoading && (
            <React.Fragment>
              <Inputs>
                <FormInput
                  label="First name"
                  placeholder="Enter first name"
                  attribute="first_name"
                  onChange={this.onInputChange}
                  value={this.state.first_name}
                  error={get(errors, "first_name")}
                  disabled={formDisabled}
                ></FormInput>

                <FormInput
                  label="Last name"
                  placeholder="Enter last name"
                  attribute="last_name"
                  onChange={this.onInputChange}
                  value={this.state.last_name}
                  error={get(errors, "last_name")}
                  disabled={formDisabled}
                ></FormInput>
              </Inputs>
              <ButtonContainer>
                <CancelButton onClick={this.close}>
                  <FiX />
                </CancelButton>
                <SaveButton
                  disabled={formDisabled}
                  loading={saveRequestStatus === REQUEST_STATUSES.PENDING}
                  onClick={
                    mode === MODES.EDIT ? this.updateUser : this.createUser
                  }
                >
                  <SaveButtonIcon />
                </SaveButton>
              </ButtonContainer>
            </React.Fragment>
          )}
        </Popover>
      </Backdrop>
    );
  }
}

UserForm.propTypes = {
  user: PropTypes.shape({
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
  }),
  userRequestStatus: PropTypes.oneOf(Object.values(REQUEST_STATUSES)),
  saveRequestStatus: PropTypes.oneOf(Object.values(REQUEST_STATUSES)),
  saving: PropTypes.bool,
  id: PropTypes.string,
  close: PropTypes.func.isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    user: get(state, `users.usersById.${ownProps.id}`),
    userRequestStatus: get(state, `requests.${GET_USERS__REQUEST}.status`),
    saveRequestStatus: get(
      state,
      `requests.${
        ownProps.id ? UPDATE_USER__REQUEST : CREATE_USER__REQUEST
      }.status`
    ),
    errors: get(
      state,
      `requests.${
        ownProps.id ? UPDATE_USER__REQUEST : CREATE_USER__REQUEST
      }.error`
    ),
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    createUser: (user) => dispatch(createAction(CREATE_USER__REQUEST, user)),
    updateUser: (user) => dispatch(createAction(UPDATE_USER__REQUEST, user)),
    resetSaveRequest: () =>
      dispatch(
        createAction(ownProps.id ? UPDATE_USER__RESET : CREATE_USER__RESET)
      ),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserForm);
