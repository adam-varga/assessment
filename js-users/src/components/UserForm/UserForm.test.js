import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";
import UserForm from "./UserForm";
import { renderWithRedux } from "../../testUtils";
import {
  createAction,
  GET_USERS__SUCCEEDED,
  CREATE_USER__SUCCEEDED,
  UPDATE_USER__SUCCEEDED,
  UPDATE_USER__REQUEST,
  CREATE_USER__REQUEST,
  GET_USERS__REQUEST,
} from "../../store/actions";
import { BrowserRouter as Router } from "react-router-dom";
import { createStore } from "redux";
import reducer from "../../store/reducers";

afterEach(cleanup);

describe("Create user form", () => {
  let renderResult;
  let mockClose = jest.fn();
  let store;

  beforeEach(() => {
    store = createStore(reducer);

    jest.spyOn(store, "dispatch");

    renderResult = renderWithRedux(
      <Router>
        <UserForm close={mockClose} />
      </Router>,
      {
        store,
      }
    );

    renderResult.store.dispatch(createAction(GET_USERS__REQUEST));
  });

  it("renders empty input fields by default regardless of whether users have been loaded", () => {
    const { getByText, queryByText, getByTestId, asFragment } = renderResult;

    expect(getByText("Create user")).toBeInTheDocument();
    expect(queryByText("Loading...")).toBeNull();

    const firstNameInput = getByTestId("input_first_name");
    const lastNameInput = getByTestId("input_last_name");

    expect(firstNameInput).toBeInTheDocument();
    expect(firstNameInput.value).toBe("");

    expect(lastNameInput).toBeInTheDocument();
    expect(lastNameInput.value).toBe("");

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls close prop function on Cancel button click", () => {
    const { getByTestId } = renderResult;

    const cancelButton = getByTestId("cancel-button");

    fireEvent(
      cancelButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(mockClose).toHaveBeenCalled();
  });

  it("dispatches CREATE_USER__REQUEST action on Save button click", () => {
    const { getByTestId } = renderResult;

    const saveButton = getByTestId("save-button");
    const firstNameInput = getByTestId("input_first_name");
    const lastNameInput = getByTestId("input_last_name");

    fireEvent.change(firstNameInput, { target: { value: "New" } });
    fireEvent.change(lastNameInput, { target: { value: "Person" } });

    store.dispatch.mockClear();

    expect(saveButton).toBeInTheDocument();

    fireEvent(
      saveButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(store.dispatch).toHaveBeenCalledWith(
      createAction(CREATE_USER__REQUEST, {
        last_name: "Person",
        first_name: "New",
        status: "active",
      })
    );
  });
});

describe("Edit user form", () => {
  let renderResult;
  let mockClose = jest.fn();
  let store;

  beforeEach(() => {
    store = createStore(reducer);

    jest.spyOn(store, "dispatch");

    renderResult = renderWithRedux(
      <Router>
        <UserForm id="1" close={mockClose} />
      </Router>,
      {
        store,
      }
    );

    renderResult.store.dispatch(createAction(GET_USERS__REQUEST));
  });

  it("renders loading screen by default", () => {
    const { getByText, queryByTestId, asFragment } = renderResult;

    expect(getByText("Edit user")).toBeInTheDocument();
    expect(getByText("Loading...")).toBeInTheDocument();
    expect(queryByTestId("input_first_name")).toBeNull();
    expect(queryByTestId("input_last_name")).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("after users request returned", () => {
    beforeEach(() => {
      const { store } = renderResult;

      store.dispatch(
        createAction(GET_USERS__SUCCEEDED, {
          usersById: {
            "1": {
              id: 1,
              last_name: "Person",
              first_name: "Test",
              status: "active",
              created_at: "2014-01-10T08:28:49.030Z",
              updated_at: "2020-06-12T14:36:40.351Z",
              url: "http://1.json",
            },
          },
          userIds: ["1"],
        })
      );
    });

    it("renders edit form prefilled with user data", () => {
      const { getByText, queryByText, getByTestId, asFragment } = renderResult;

      expect(getByText("Edit user")).toBeInTheDocument();
      expect(queryByText("Loading...")).toBeNull();

      const firstNameInput = getByTestId("input_first_name");
      const lastNameInput = getByTestId("input_last_name");

      expect(firstNameInput).toBeInTheDocument();
      expect(firstNameInput.value).toBe("Test");

      expect(lastNameInput).toBeInTheDocument();
      expect(lastNameInput.value).toBe("Person");

      expect(asFragment()).toMatchSnapshot();
    });

    it("calls close prop function on Cancel button click", () => {
      const { getByTestId } = renderResult;

      const cancelButton = getByTestId("cancel-button");

      fireEvent(
        cancelButton,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(mockClose).toHaveBeenCalled();
    });

    it("dispatches UPDATE_USER__REQUEST action on Save button click", () => {
      const { getByTestId } = renderResult;

      const saveButton = getByTestId("save-button");
      const firstNameInput = getByTestId("input_first_name");
      const lastNameInput = getByTestId("input_last_name");

      fireEvent.change(firstNameInput, { target: { value: "Modified" } });
      fireEvent.change(lastNameInput, { target: { value: "Human" } });

      store.dispatch.mockClear();

      expect(saveButton).toBeInTheDocument();

      fireEvent(
        saveButton,
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
        })
      );

      expect(store.dispatch).toHaveBeenCalledWith(
        createAction(UPDATE_USER__REQUEST, {
          id: 1,
          last_name: "Human",
          first_name: "Modified",
          status: "active",
          created_at: "2014-01-10T08:28:49.030Z",
          updated_at: "2020-06-12T14:36:40.351Z",
          url: "http://1.json",
        })
      );
    });
  });
});
