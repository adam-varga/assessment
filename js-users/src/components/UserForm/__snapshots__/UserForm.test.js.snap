// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create user form renders empty input fields by default regardless of whether users have been loaded 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw hQgxMr"
  >
    <div
      class="sc-AxhCb jnKDwQ"
    >
      <h2
        class="sc-AxhUy dXZGiX"
      >
        Create user
      </h2>
      <div
        class="sc-AxgMl gHtrlA"
      >
        <div
          class="sc-AxheI eZozzG"
        >
          <label
            class="sc-Axmtr ezRnLP"
          >
            <span>
              First name
            </span>
            <input
              class="sc-AxmLO dmsnnn"
              data-testid="input_first_name"
              placeholder="Enter first name"
              value=""
            />
          </label>
          <p
            class="sc-fzozJi fVqnZq"
          />
        </div>
        <div
          class="sc-AxheI eZozzG"
        >
          <label
            class="sc-Axmtr ezRnLP"
          >
            <span>
              Last name
            </span>
            <input
              class="sc-AxmLO dmsnnn"
              data-testid="input_last_name"
              placeholder="Enter last name"
              value=""
            />
          </label>
          <p
            class="sc-fzozJi fVqnZq"
          />
        </div>
      </div>
      <div
        class="sc-fzoLsD hNSacB"
      >
        <button
          class="sc-AxjAm sc-fzplWN enEGyk"
          data-testid="cancel-button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
        <button
          class="sc-AxjAm sc-fzpans grbCNx"
          data-testid="save-button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z"
            />
            <polyline
              points="17 21 17 13 7 13 7 21"
            />
            <polyline
              points="7 3 7 8 15 8"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Edit user form after users request returned renders edit form prefilled with user data 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw hQgxMr"
  >
    <div
      class="sc-AxhCb jnKDwQ"
    >
      <h2
        class="sc-AxhUy dXZGiX"
      >
        Edit user
      </h2>
      <div
        class="sc-AxgMl gHtrlA"
      >
        <div
          class="sc-AxheI eZozzG"
        >
          <label
            class="sc-Axmtr ezRnLP"
          >
            <span>
              First name
            </span>
            <input
              class="sc-AxmLO dmsnnn"
              data-testid="input_first_name"
              placeholder="Enter first name"
              value="Test"
            />
          </label>
          <p
            class="sc-fzozJi fVqnZq"
          />
        </div>
        <div
          class="sc-AxheI eZozzG"
        >
          <label
            class="sc-Axmtr ezRnLP"
          >
            <span>
              Last name
            </span>
            <input
              class="sc-AxmLO dmsnnn"
              data-testid="input_last_name"
              placeholder="Enter last name"
              value="Person"
            />
          </label>
          <p
            class="sc-fzozJi fVqnZq"
          />
        </div>
      </div>
      <div
        class="sc-fzoLsD hNSacB"
      >
        <button
          class="sc-AxjAm sc-fzplWN enEGyk"
          data-testid="cancel-button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
        <button
          class="sc-AxjAm sc-fzpans grbCNx"
          data-testid="save-button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z"
            />
            <polyline
              points="17 21 17 13 7 13 7 21"
            />
            <polyline
              points="7 3 7 8 15 8"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Edit user form renders loading screen by default 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw hQgxMr"
  >
    <div
      class="sc-AxhCb jnKDwQ"
    >
      <h2
        class="sc-AxhUy dXZGiX"
      >
        Edit user
      </h2>
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
